packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,991 @@
+    ],
+  },
+  defaultOptions: [
+    {
+      considerOtherOverloads: true,
+      considerBaseClass: true,
+      considerImplementedInterfaces: true,
+      allowReturnPromiseIfTryCatch: true,
+      allowReturnUndefined: true,
+      allowReturnNull: true,
+    },
+  ],
+
+  create(context, [options]) {
+    const sourceCode = context.sourceCode;
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      'CallExpression, NewExpression': (
+        node: TSESTree.CallExpression | TSESTree.NewExpression,
+      ): void => {
+        checkFunctionCallNode(node);
+      },
+      JSXAttribute: (node): void => {
+        if (
+          node.value?.type === AST_NODE_TYPES.JSXExpressionContainer &&
+          node.value.expression.type !== AST_NODE_TYPES.JSXEmptyExpression
+        ) {
+          const attrName = sourceCode.getText(node.name);
+          const elemName = sourceCode.getText(node.parent.name);
+          checkExpressionNode(node.value.expression, 'Attr', {
+            attrName,
+            elemName,
+          });
+        }
+      },
+      VariableDeclarator: (node): void => {
+        if (node.init != null) {
+          const varName = util.getNameFromExpression(sourceCode, node.id);
+          if (varName != null) {
+            checkExpressionNode(node.init, 'Var', { varName });
+          } else {
+            checkExpressionNode(node.init, 'Other');
+          }
+        }
+      },
+      AssignmentExpression: (node): void => {
+        if (['=', '||=', '&&=', '??='].includes(node.operator)) {
+          const varName = util.getNameFromExpression(sourceCode, node.left);
+          if (varName != null) {
+            checkExpressionNode(node.right, 'Var', { varName });
+          } else {
+            checkExpressionNode(node.right, 'Other');
+          }
+        }
+      },
+      ObjectExpression: (node): void => {
+        for (const propNode of node.properties) {
+          if (propNode.type !== AST_NODE_TYPES.SpreadElement) {
+            checkObjectPropertyNode(propNode);
+          }
+        }
+      },
+      ArrayExpression: (node): void => {
+        for (const elemNode of node.elements) {
+          if (
+            elemNode != null &&
+            elemNode.type !== AST_NODE_TYPES.SpreadElement
+          ) {
+            checkExpressionNode(elemNode, 'Other');
+          }
+        }
+      },
+      ArrowFunctionExpression: (node): void => {
+        if (node.body.type !== AST_NODE_TYPES.BlockStatement) {
+          checkExpressionNode(node.body, 'Return');
+        }
+      },
+      ReturnStatement: (node): void => {
+        if (node.argument != null) {
+          checkExpressionNode(node.argument, 'Return');
+        }
+      },
+      PropertyDefinition: (node): void => {
+        checkClassPropertyNode(node);
+      },
+      MethodDefinition: (node): void => {
+        checkClassMethodNode(node);
+      },
```

Comment:
I'd say let's wait on that, yeah. Maybe in a future PR.

Reply:
Including for `no-misused-promises`?

Reply:
Separate from this PR for sure, but I meant do we want to postpone that indefinitely until we have a better idea of what we're going to do with `no-misused-promises`?

Reply:
Oh, gotcha - no I think that's a different thing. Whether a rule properly checks index signatures is a _bug_ scoped _with the rule_. High-level direction discussions like this might change what rule(s) exist or where they look, but don't really impact small bugs like that.